### PR TITLE
Generate RSS manifests

### PR DESCRIPTION
**Pull request by hosamaly**
Clean up and test the generation of RSS manifests, and make sure XML is escaped properly.

### CODE REVIEW

File: src/main/scala/com/gu/kindlegen/Article.scala [hunk 1]
```diff
@@ -19,7 +19,8 @@ object Article {
   def getBodyBlocks(content: Content): Seq[String] = {
     val blocks = content.blocks
     val bodyBlocks = blocks.flatMap(_.body).getOrElse(Nil)
-    // TODO what if bodyBlocks.length < blocks.totalBodyBlocks ?
+      .ensuring(bodies => blocks.flatMap(_.totalBodyBlocks).forall(_ == bodies.length))
```
Comment (tomrf1) [inline src/main/scala/com/gu/kindlegen/Article.scala]: I don't think this check is necessary - capi should always return all blocks if requested

File: src/main/scala/com/gu/kindlegen/Article.scala [hunk 2]
```diff
@@ -40,7 +41,7 @@ object Article {
   def apply(content: Content, newspaperDate: CapiDateTime, fields: ContentFields, sectionTag: Tag): Article = {
     Article(
       section = Section(sectionTag),
-      newspaperPageNumber = fields.newspaperPageNumber.getOrElse(0),
+      newspaperPageNumber = fields.newspaperPageNumber.getOrElse(Int.MaxValue),  // move to the end of the section
```
Comment (tomrf1) [inline src/main/scala/com/gu/kindlegen/Article.scala]: how is this field used? I'm guessing it's not included in the output

File: src/main/scala/com/gu/kindlegen/Rss.scala [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+      </channel>
+    </rss>
+}
+
+object RssManifest {
+  private final val Encoding = "UTF-8"
+
+  private val formatter = DateTimeFormatter
+    .ofPattern("EEE, dd MMM yyyy HH:mm:ss Z")  // Tue, 01 May 2018 17:10:54 +0000
+    .withZone(ZoneId.of("UTC"))
```
Comment (regiskuckaertz) [inline src/main/scala/com/gu/kindlegen/Rss.scala]: There's a constant if you prefer: `ZoneOffset.UTC`

File: src/main/scala/com/gu/kindlegen/Article.scala (hunk 1 above)
Comment (hosamaly) [reply src/main/scala/com/gu/kindlegen/Article.scala]: I guess this is a question of how defensive / trusting this code should be. This assertion may not guard against CAPI as much as it defends against an incorrect query where `show-blocks` is missing or set to an undesired value (e.g. `show-blocks=main`).

Do you think this is the wrong place to have this check?

File: src/main/scala/com/gu/kindlegen/Article.scala (hunk 2 above)
Comment (hosamaly) [reply src/main/scala/com/gu/kindlegen/Article.scala]: It's used in `BookSection` to [sort sections](https://github.com/guardian/kindle-gen/blob/a9d147154bc9b1b116568df271e4b7c464ae78b5/src/main/scala/com/gu/kindlegen/BookSection.scala#L30) and to [sort articles in a section](https://github.com/guardian/kindle-gen/blob/a9d147154bc9b1b116568df271e4b7c464ae78b5/src/main/scala/com/gu/kindlegen/BookSection.scala#L43).

File: src/main/scala/com/gu/kindlegen/Rss.scala (hunk 3 above)
Comment (hosamaly) [reply src/main/scala/com/gu/kindlegen/Rss.scala]: Thanks. I hadn't realised that `ZoneOffset extends ZoneId`. I'll change it.

File: src/main/scala/com/gu/kindlegen/BookSection.scala [hunk 4]
```diff
@@ -13,8 +17,9 @@ case class BookSection(section: Section, articles: Seq[Article]) extends Linkabl
   def id: String = section.id
   def title: String = section.title
   def link: Link = section.link
+  lazy val publicationDate: LocalDate = articles.map(_.pubDate).minBy(_.dateTime).toOffsetDateTime.toLocalDate
 
-  def publicationDate = articles.map(_.pubDate).minBy(_.dateTime)
+  def withLink(newLink: Link): BookSection = copy(section = section.copy(link = newLink))
```
Comment (regiskuckaertz) [inline src/main/scala/com/gu/kindlegen/BookSection.scala]: It seems you are using this solely for the purpose of collecting local file names at the end of your compilation. You end up doing copies of your value objects all the way down ([here](https://github.com/guardian/kindle-gen/blob/a9d147154bc9b1b116568df271e4b7c464ae78b5/src/main/scala/com/gu/kindlegen/KindleGenerator.scala#L64), [here](https://github.com/guardian/kindle-gen/blob/a9d147154bc9b1b116568df271e4b7c464ae78b5/src/main/scala/com/gu/kindlegen/KindleGenerator.scala#L75) and [here](https://github.com/guardian/kindle-gen/blob/a9d147154bc9b1b116568df271e4b7c464ae78b5/src/main/scala/com/gu/kindlegen/KindleGenerator.scala#L97)). That may create a lot of GC pressure as objects are created and consumed immediately -- I don't know how good Scala is at structural sharing. Getting around that is not very hard, so if you feel like it I can give you the gist of it.

File: src/main/scala/com/gu/kindlegen/Article.scala (hunk 1 above)
Comment (tomrf1) [reply src/main/scala/com/gu/kindlegen/Article.scala]: the `totalBodyBlocks` field is only added if the client has requested body blocks, so this test is unnecessary - https://github.com/guardian/content-api/blob/c93ebc8e55b9270542a66d9a66ce33521e6984b5/concierge/src/main/scala/com.gu.contentapi.concierge/processors/ContentProcessor.scala#L179

Comment (hosamaly) [reply src/main/scala/com/gu/kindlegen/Article.scala]: I removed the check in 0a50ec3.

**Pull request merged**